### PR TITLE
feat(eip8130): EIP-8130 rpc extensions

### PR DESCRIPTION
**Pull request by lukasrosario**
## Summary

Extends `eth_getTransactionCount` with an optional EIP-8130 `nonce_key` (uint256) parameter as a third positional argument, enabling RPC clients to query 2D nonce channels from the Nonce Manager precompile.

Behavior by `nonce_key`:
- **omitted** → existing protocol-nonce behavior (backwards-compatible).
- **`0`** → protocol nonce (collapses with omitted post-Cobalt, per EIP-8130's reservation of key 0).
- **non-zero `k`** → 2D channel nonce from `nonces[address][k]` in the precompile's storage.
- **`NONCE_KEY_MAX`** → `INVALID_PARAMS` (expiring-nonce channel has no per-channel counter).
- **any value pre-Cobalt** → `INVALID_PARAMS` (gate mirrors the txpool's `TxTypeNotSupported` rejection of EIP-8130 txs pre-activation).

## Design

**Single owner for `eth_getTransactionCount`**. Both the existing flashblocks override and a new standalone override register the same method via jsonrpsee's overwrite-semantics `replace_configured`. Resolved by self-gating at the node-assembly site in `standard_node.rs`: `flashblocks_config.is_some()` determines which extension installs, so exactly one owns the method.

| Node config | Owner | Behavior on pending |
|---|---|---|
| Flashblocks enabled | flashblocks override | protocol nonce: `canon + flashblock_delta`; channel: precompile slot read with flashblocks `StateOverride` layered on canonical state |
| Flashblocks disabled | standalone `Eip8130EthApiExt` | standard reth resolution + raw precompile slot read |

**Channel read avoids `eth_call`**. `NonceManagerStorage::nonce_slot(addr, key)` (new) derives the slot via the same Solidity-compatible math the precompile's typed handler chain uses internally, and the RPC reader does a single `StateProvider::storage` lookup. This is ~10× faster than an `eth_call` to the precompile's `getNonce` while keeping layout ownership inside `base-common-precompiles` (the precompile crate is the single source of truth; off-chain readers go through the typed helper).

**Cobalt gate on `nonce_key.is_some()`**. Gates the whole EIP-8130 RPC surface on Cobalt activation, including explicit `Some(0)`. A client probing for EIP-8130 support by sending `nonce_key = 0` gets an unambiguous "feature not active" pre-Cobalt instead of silently delegating. Block-of-query timestamp drives the gate.

## Changes by crate

- **`base-common-precompiles`**: `NonceManagerStorage::nonce_slot()` static helper + `NONCES_BASE_SLOT` const.
- **`base-execution-eip8130-rpc`** *(new)*: `ChannelNonceReader` (slot derivation + override-aware read), `Eip8130CobaltGate` (fork gate), `Eip8130EthApiOverride` trait + `Eip8130EthApiExt<Eth>` impl (standalone non-flashblocks override).
- **`base-execution-eip8130-rpc-node`** *(new)*: `Eip8130RpcExtension` (`BaseNodeExtension` wiring with self-gating bool).
- **`base-flashblocks`**: flashblocks's `EthApiOverride::get_transaction_count` extended with `nonce_key: Option<U256>` parameter; routes through `ChannelNonceReader` for the non-zero case, layering `pending_blocks.get_state_overrides()` for the pending path.
- **`base-execution-cli`**: installs `Eip8130RpcExtension` after `FlashblocksExtension`, gated on `!flashblocks_config.is_some()`.

## Wire compatibility

JSON-RPC change is fully backwards-compatible. `jsonrpsee` deserializes a missing trailing `Option<U256>` as `None`, so existing 2-arg `eth_getTransactionCount(address, blockNumber)` calls continue to behave identically (route through the legacy path with no gate).

## Test plan

- [x] Unit tests for `nonce_slot()` (4): handler-chain consistency, end-to-end sload-and-decode, distinct slots per (account, key), protocol-nonce rejection.
- [x] Unit tests for override lookup + u64 decode (10): all override shapes, sentinel/edge values, high-bit ignoring.
- [x] `cargo check -p base-execution-cli --tests` clean. Verifies the whole assembly chain (cli + flashblocks-node + the 2 new crates) compiles together.
- [x] `cargo test -p base-flashblocks --lib` (111 tests). No regression on flashblocks's existing override behavior.
- [ ] **Node-level integration tests deferred.** Across the existing EIP-8130 implementation (validator, txpool admission, precompile dispatch, span batches) there are zero integration tests in any `tests/` directory. The whole feature relies on colocated unit tests. Building a Cobalt-activated test harness + EIP-8130 tx-submission helpers is meaningful work that benefits the entire feature, not just our RPC slice. Tracked as a follow-up; coverage profile matches the rest of EIP-8130 in the meantime.

## Open items for follow-up

1. **EIP-8130 integration test infrastructure**. Cobalt-activated genesis builder, `FlashblocksHarness` chain-spec parameterization (or new non-flashblocks harness), EIP-8130 tx-submission helpers, raw 3-arg `eth_getTransactionCount` plumbing. Reusable across the whole feature.
2. **`NONCE_KEY_MAX` RPC semantics**. Using `INVALID_PARAMS` ("no per-channel counter") based on the spec's framing of NONCE_KEY_MAX as the expiring-nonce sentinel.
3. **`state` vs `state_diff` precedence** in `override_for_slot`. Geth's `eth_call` overrides extension doesn't formally spec precedence when both are set. Current behavior: `state` (full replacement) takes precedence over `state_diff` (merge). No test asserts this (intentionally, since it's underspecified).

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>✅ Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
✅
<code>1/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

**Comment by Sertug17** [comment]
Same pattern as the nonce_reader .expect() that the bot flagged the `try_into().expect("8 bytes")` in `decode_channel_nonce` is guaranteed to succeed since the slice is always 8 bytes, so it won't panic in practice. But if someone refactors the slice bounds or the slot layout later, this turns into an RPC panic the same way. Could just use `u64::from_be_bytes(bytes[24..32].try_into().unwrap())` since `unwrap()` on a known-infallible conversion communicates intent better than `expect()` with a message that's about the current implementation, not the invariant.

File: crates/execution/eip8130-rpc/src/nonce_reader.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        Eth: FullEthApi<NetworkTypes = Base> + Send + Sync + 'static,
+        ErrorObjectOwned: From<Eth::Error>,
+    {
+        // Protocol nonce. Lives in account state, not the precompile.
+        // Delegate to the standard `eth_getTransactionCount` resolution path.
+        if nonce_key == U256::ZERO {
+            return EthState::transaction_count(eth_api, address, Some(block_id))
+                .await
+                .map_err(Into::into);
+        }
```
[the remark this answers is not in the document]
Comment (haardikk21) [reply crates/execution/eip8130-rpc/src/nonce_reader.rs:62]: This makes sense imo. Would prefer having these two things be protected via regression tests behaviourally rather than the guards.

File: crates/execution/eip8130-rpc/src/nonce_reader.rs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    ///
+    /// Widening to `U256` matches the return type of
+    /// [`EthState::transaction_count`] so all three `nonce_key` branches of
+    /// [`Self::read`] return the same shape.
+    pub fn decode_channel_nonce(slot_value: U256) -> U256 {
+        slot_value & U256::from(u64::MAX)
+    }
+}
+
+const fn b256_to_u256(value: B256) -> U256 {
```
Comment (haardikk21) [inline crates/execution/eip8130-rpc/src/nonce_reader.rs]: Can we just inline this where necessary? top level functions like this are unnecessary indirection

File: crates/execution/flashblocks/src/rpc/eth.rs [hunk 3]
```diff
@@ -264,13 +270,51 @@ where
...
         );
 
         let block_id = block_number.unwrap_or_default();
+
+        // EIP-8130 surface. Any `nonce_key` (including `Some(0)`) is the
+        // client using the EIP-8130 RPC extension, so we gate the whole
+        // surface on Cobalt. Mirrors the txpool's `TxTypeNotSupported`
+        // rejection of EIP-8130 transactions pre-activation.
+        if let Some(key) = nonce_key {
+            Eip8130CobaltGate::check(&self.eth_api, block_id)?;
```
[the remark this answers is not in the document]
Comment (haardikk21) [reply crates/execution/flashblocks/src/rpc/eth.rs:288]: Agreed, let's not introduce sync reads in the "default" path of how folks do things. This is hot path code used very heavily by top customers. Let's gate the cobalt check on nonce key being non-zero only

**Review by haardikk21 — APPROVED** [review]
